signal-hook-mio/tests/mio.rs

```diff
@@ -0,0 +1,111 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::thread;
+use std::time::Duration;
+
+use mio::{Events, Interest, Poll, Token};
+
+use signal_hook_mio::Signals;
+
+use libc::c_int;
+
+#[test]
+fn mio_wakeup() {
+    let mut signals = Signals::new(&[signal_hook::SIGUSR1]).unwrap();
+    let mut poll = Poll::new().unwrap();
+    let token = Token(0);
+    poll.registry()
+        .register(&mut signals, token, Interest::READABLE)
+        .unwrap();
+
+    let mut events = Events::with_capacity(10);
+
+    // The self pipe shouldn't be readable yet
+    poll.poll(&mut events, Some(Duration::from_secs(0)))
+        .unwrap();
+    assert!(events.is_empty());
+
+    unsafe { libc::raise(signal_hook::SIGUSR1) };
+    poll.poll(&mut events, Some(Duration::from_secs(10)))
+        .unwrap();
+    let event = events.iter().next().unwrap();
+
+    assert!(event.is_readable());
+    assert_eq!(token, event.token());
+    let sig = signals.pending().next().unwrap();
+    assert_eq!(signal_hook::SIGUSR1, sig);
+
+    // The self pipe shouldn't be readable after consuming signals
+    poll.poll(&mut events, Some(Duration::from_secs(0)))
+        .unwrap();
+    assert!(events.is_empty());
```

Comment:
I noticed some of the tests are a bit flaky. Sometimes some tests fail but it weren't always the same tests. But tests in the same file running in parallel may explain that.

Maybe we could also run the test with only one thread instead. This is possible by appending `-- --test-threads=1` at the end of the cargo test command. This would be easier than introducing global mutex's in a bunch of tests but affects all tests whether they need it or not.

Reply:
That's possible, but has the downside of every contributor that happens to just type `cargo test` having to be aware of that.

Reply:
You're right, for the CI script this would be OK but it is a real burden for contributor. So `--test-threads=1` isn't a good solution.

Having one file per test case would also mitigate this issue because then each test case would become a separate executable and run in its own process in isolation from the other test cases. But I'm not sure if it is good to rely on this behavior of Rust / Cargo is fixed or if it could change in the future.

Then there is also [serial_test crate](https://crates.io/crates/serial_test). This create requires Rust >= 1.39. But this should be OK because the 1.26 and 1.31 CI builds do not execute the test cases.

Reply:
> But I'm not sure if it is good to rely on this behavior of Rust / Cargo is fixed or if it could change in the future.

I think changing that would be breaking a lot of code, so this won't change. But it also feels like there would be a lot of boilerplate when placing each one into a separate file.

The serial_test crate is fine too. I think it just does the mutex thing too.

Reply:
A use the serial_test create now in the adapter crates and the iterator module tests. I didn't touch the other tests (pipe, flag and cleanup) yet.